src/main.rs

```diff
@@ -41,20 +41,40 @@ async fn main() -> anyhow::Result<()> {
             .map(|streamer_message| handle_streamer_message(streamer_message, &pool))
             .buffer_unordered(1usize);
 
-        while let Some(_handle_message) = handlers.next().await {}
+        let mut time_now = std::time::Instant::now();
+        while let Some(handle_message) = handlers.next().await {
+            match handle_message {
+                Ok(block_height) => {
+                    let elapsed = time_now.elapsed();
+                    tracing::info!(
+                        target: LOGGING_PREFIX,
+                        "Elapsed time spent on block {}: {:.3?}",
+                        block_height,
+                        elapsed
+                    );
+                    time_now = std::time::Instant::now();
+                }
+                Err(e) => {
+                    tracing::error!(target: LOGGING_PREFIX, "Stop indexing due to {}", e);
+                    // we do not catch this error anywhere, this thread is just stopped with error,
+                    // main thread continues serving metrics
```

Comment:
@roshaans @morgsmccauley we need to add this `tracing::error!` to all of our indexers, that's the only way not to ignore the error that we meet during running indexer

Reply:
I'll add to explorer indexer 👍🏽

Reply:
Actually, I have a some questions about this:

1. Is bailing really the right thing to do here? What is the point of just killing this thread and allowing metrics to continue? I suppose we would see the metrics decline and can therefore respond, but wouldn't be better to catch this error in the main thread and kill the entire process?
2. What kind of errors are we catching here? Are the error recoverable? In which case we should not bail at all and instead allow the stream to continue processing?

Reply:
If we fail and come here, it 100% means the error is non-recoverable.
I don't think we should kill the metrics, we can continue serve them, showing that everything is not OK. (I am 80% sure with this decision)
It would be nice if you collect more opinions on how it's better to handle such cases

Reply:
I think having no metrics available at all (i.e. the service is dead) would be the same as having a metrics server running, but no new data available. So in either case we would see the metrics decline in Grafana.

Additionally, killing the entire process will mean Cloud Run will restart it, and maybe restarting is enough to resolve the issue?

@ecp88 Maybe you have some input on this?

Reply:
> Additionally, killing the entire process will mean Cloud Run will restart it, and maybe restarting is enough to resolve the issue?

Ahhh that's the great point you've mentioned.
Restarting will NOT help for sure in most of the cases. If it may help, it means we should have handled the error at the very beginning, while we are saying the error was non-recoverable.

The only case I can imagine when silent restart helps, is when something is down (e.g. S3 servers), we tried to wait for the answer for few minutes, then gave up. The process is restarted again and again. After an hour of this work, S3 is responsive again, and everything is fixed. 
But I hope it's a rare scenario and we can rely on S3.

More common scenario (maybe not more common, but the scenario that we have to handle carefully) is when we see data that we can't work with. You can restart the process million times, the data will not change.
The example is here: https://github.com/near/near-lake-indexer/issues/47#issuecomment-1346480729
The best that we can do is to discuss what's going on and how we need to live with it. 
And we usually need to make the decision manually. 
In the example, the simplest decision was to create the issue describing all the pain we see, and restart the indexer from the next block after the problematic one. But it's not always the behavior that we want

Reply:
@khorolets I want you in this discussion

Reply:
IMO @telezhnaya  is right when she wants Grafana to know that everything is not OK if the error is non-recoverable. Also, @morgsmccauley is right in saying that the killed process should not respond with metrics and it's the same thing.

I think it's OK to kill the entire process **if we can guarantee Grafana or whatever receives the signal that something went wrong and notifies us**. If the auto-restart feature of Cloud Run doesn't brake the process of understanding there is a problem we should proceed with the @morgsmccauley suggestion (terminate the process in case of non-recoverable errors).

P.S. I just want to remind you that our strategy of not killing the process is related to the stuff built on top of Indexer Framework (nearcore) and the main reasoning was to avoid dropping from the sync for the node. Since this is Lake-based indexer we don't have this reason anymore.

Reply:
Did some playing round with Prometheus locally, tried killing indexer after a while to see what happens.

For BPS, the metrics just stop at the last time period, rather than dropping to `0`. This does make sense as no new metrics are coming in, but I guess it would appear as we still have BPS while the indexer is actually dead. There's probably some way to make this absence of values more obvious.

A more straight forward approach is to use the `up()` function to check whether the service is running via Grafana. This will return `0` when Prometheus can't reach the service so is basically what we want.

Reply:
Did some playing round with Prometheus locally, tried killing indexer after a while to see what happens.

BPS drops to `0` after indexer is killed, so we'd be able to alert off this to know when the indexer has died.

Another more straight forward approach is to use the `up()` function to check whether the service is running via Grafana. This will return `0` when Prometheus can't reach the service.